crates/admin/src/schema_registry/mod.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+pub mod error;
+mod updater;
+
+use crate::schema_registry::error::{SchemaError, SchemaRegistryError};
+use crate::schema_registry::updater::SchemaUpdater;
+use http::Uri;
+use restate_core::metadata_store::MetadataStoreClient;
+use restate_core::{metadata, MetadataWriter};
+use restate_schema::SchemaInformation;
+use restate_schema_api::component::{ComponentMetadata, ComponentMetadataResolver};
+use restate_schema_api::deployment::{DeliveryOptions, DeploymentMetadata, DeploymentResolver};
+use restate_schema_api::subscription::{Subscription, SubscriptionResolver, SubscriptionValidator};
+use restate_service_client::Endpoint;
+use restate_service_protocol::discovery::{ComponentDiscovery, DiscoverEndpoint};
+use restate_types::identifiers::{DeploymentId, SubscriptionId};
+use restate_types::metadata_store::keys::SCHEMA_INFORMATION_KEY;
+use std::collections::HashMap;
+use std::ops::Deref;
+use tracing::subscriber::NoSubscriber;
+
+/// Whether to force the registration of an existing endpoint or not
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Force {
+    Yes,
+    No,
+}
+
+impl Force {
+    pub fn force_enabled(&self) -> bool {
+        *self == Self::Yes
+    }
+}
+
+/// Whether to apply the changes or not
+#[derive(Clone, Default, PartialEq, Eq, Debug)]
+pub enum ApplyMode {
+    DryRun,
+    #[default]
+    Apply,
+}
+
+impl ApplyMode {
+    pub fn should_apply(&self) -> bool {
+        *self == Self::Apply
+    }
+}
+
+/// Responsible for updating the registered schema information. This includes the discovery of
+/// new deployments.
+#[derive(Clone)]
+pub(crate) struct SchemaRegistry<V> {
+    metadata_store_client: MetadataStoreClient,
+    metadata_writer: MetadataWriter,
+    component_discovery: ComponentDiscovery,
+    subscription_validator: V,
+}
+
+impl<V> SchemaRegistry<V> {
+    pub(crate) fn new(
+        metadata_store_client: MetadataStoreClient,
+        metadata_writer: MetadataWriter,
+        component_discovery: ComponentDiscovery,
+        subscription_validator: V,
+    ) -> Self {
+        Self {
+            metadata_writer,
+            metadata_store_client,
+            component_discovery,
+            subscription_validator,
+        }
+    }
+
```

Comment:
How about you add the "read" methods here as well that are used by the rest_api handlers? Makes it more lean and self contained.

Reply:
Good idea. Will move them here.